### ISSUE TITLE
Cluster init/node join has often problems with stale node daemons

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#201]
Originally reported of Google Code with ID 145.

```
Many times, users have run into issues where cluster init cannot complete due to old node daemons; alternatively, node joins fail also due to existing node daemons.

We should simply check on cluster init for running node daemon, and on node join for remotely open noded port.
```
Originally added on 2011-02-24 09:05:28 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
This happens to me as well when running gnt-cluster init on gentoo with ganeti 2.4.2, here is the debug output:

vm0 ~ # gnt-cluster init --debug --secondary-ip=$drbd_ip --vg-name=$vgname --master-netdev=$interfacename --enabled-hypervisors=$hypervisors --prealloc-wipe-disks=no $clustername
2011-08-02 15:23:46,953: gnt-cluster init pid=2332 cli:1927 INFO run with arguments '--debug --secondary-ip=10.1.0.23 --vg-name=ganeti --master-netdev=lan --enabled-hypervisors=kvm --prealloc-wipe-disks=no ganeti.hq.*deleted*.hu'
2011-08-02 15:23:46,953: gnt-cluster init pid=2332 rpc:93 INFO Using PycURL libcurl/7.21.4 OpenSSL/1.0.0d zlib/1.2.5
2011-08-02 15:23:49,961: gnt-cluster init pid=2332 process:195 DEBUG RunCmd vgs --noheadings --units m --nosuffix -o name,size
2011-08-02 15:23:50,006: gnt-cluster init pid=2332 process:195 DEBUG RunCmd ip link show dev lan
2011-08-02 15:23:50,077: gnt-cluster init pid=2332 process:195 DEBUG RunCmd ssh-keygen -t dsa -f /root/.ssh/id_dsa -q -N ''
2011-08-02 15:23:51,662: gnt-cluster init pid=2332 bootstrap:124 DEBUG Generating new cluster certificate at /var/lib/ganeti/server.pem
2011-08-02 15:23:51,853: gnt-cluster init pid=2332 bootstrap:129 DEBUG Writing new confd HMAC key to /var/lib/ganeti/hmac.key
2011-08-02 15:23:51,902: gnt-cluster init pid=2332 bootstrap:144 DEBUG Generating new RAPI certificate at /var/lib/ganeti/rapi.pem
2011-08-02 15:23:52,542: gnt-cluster init pid=2332 bootstrap:153 DEBUG Generating new cluster domain secret at /var/lib/ganeti/cluster-domain-secret
2011-08-02 15:23:52,584: gnt-cluster init pid=2332 process:195 DEBUG RunCmd /usr/lib64/ganeti/daemon-util start ganeti-noded
2011-08-02 15:23:52,829: gnt-cluster init pid=2332 client:335 DEBUG Starting request <ganeti.http.client.HttpClientRequest 10.0.0.23:1811 PUT /version at 0x24e0b10>
2011-08-02 15:23:52,829: gnt-cluster init pid=2332 client:320 DEBUG Created new client <ganeti.http.client._PooledHttpClient id=10.0.0.23/1811 lastuse=0 <ganeti.http.client._HttpClient object at 0x24e0810> at 0x24eaf38>
2011-08-02 15:23:52,864: gnt-cluster init pid=2332 client:232 DEBUG Request <ganeti.http.client.HttpClientRequest 10.0.0.23:1811 PUT /version at 0x24e0b10> finished, errmsg=None
2011-08-02 15:23:52,865: gnt-cluster init pid=2332 client:350 DEBUG Returning client <ganeti.http.client._PooledHttpClient id=10.0.0.23/1811 lastuse=1 <ganeti.http.client._HttpClient object at 0x24e0810> at 0x24eaf38> to pool
2011-08-02 15:23:52,865: gnt-cluster init pid=2332 bootstrap:446 DEBUG Starting daemons
2011-08-02 15:23:52,866: gnt-cluster init pid=2332 process:195 DEBUG RunCmd /usr/lib64/ganeti/daemon-util start-all
2011-08-02 15:23:52,964: gnt-cluster init pid=2332 process:116 DEBUG Command '/usr/lib64/ganeti/daemon-util start-all' failed (exited with exit code 1); output:  * start-stop-daemon: /usr/sbin/ganeti-noded is already running
exit code 1

2011-08-02 15:23:52,965: gnt-cluster init pid=2332 cli:1936 ERROR Error during command processing
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/ganeti/cli.py", line 1932, in GenericMain
    result = func(options, args)
  File "/usr/lib64/python2.7/site-packages/ganeti/rpc.py", line 176, in wrapper
    return fn(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/ganeti/client/gnt_cluster.py", line 146, in InitCluster
    prealloc_wipe_disks=opts.prealloc_wipe_disks,
  File "/usr/lib64/python2.7/site-packages/ganeti/bootstrap.py", line 451, in InitCluster
    (result.cmd, result.exit_code, result.output))
OpExecError: Could not start daemons, command /usr/lib64/ganeti/daemon-util start-all had exitcode 1 and error  * start-stop-daemon: /usr/sbin/ganeti-noded is already running
exit code 1

Failure: command execution error:
Could not start daemons, command /usr/lib64/ganeti/daemon-util start-all had exitcode 1 and error  * start-stop-daemon: /usr/sbin/ganeti-noded is already running
exit code 1

Please note, that before running this command there were no ganeti daemons running. Init is trying to start ganeti-noded twice, first at 15:23:52,584 and later at 5:23:52,866.
```
Originally added on 2011-08-02 14:28:35 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2012-07-19 13:07:12 +0000 UTC.

Added Labels: Type-Defect

**Comment by Ganeti-Issues-Migrator** [comment]
```
I just encountered the same thing on a recent deployment. Out of curiosity, what version of openrc do you have installed? I'm trying to debug this myself.

Gentoo ships a their own variant of start-stop-daemon and unfortunately sometimes they don't work the same way that the Debian does. I may need to write a patch as a work around (similar to what the RedHat patch does).
```
Originally added on 2012-07-25 20:25:45 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I created a nasty work around hack that can probably be done in many other different (better) ways. It basically checks to see if the daemon in question is already running using pgrep and skips executing start-stop-daemon if it does see it. 

I was able to at least get past the major blocker. Please let me know if you have a better solution for this. This is only needed on Gentoo hosts as far as I can tell.
```
Originally added on 2012-07-25 23:29:12 +0000 UTC.

Attachments: 
[0001-Check-if-process-is-already-running-and-skip-if-so.patch](https://storage.googleapis.com/google-code-attachments/ganeti/issue-145/comment-4/0001-Check-if-process-is-already-running-and-skip-if-so.patch)

**Comment by Ganeti-Issues-Migrator** [comment]
```
Assigning to myself as I did work to improve the situation on node-add by returning errors if a certificate doesn't match and stopping running daemons. Cluster initialization is now left to be done.
```
Originally added on 2012-12-18 17:24:17 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Ramereth, I know it's abit late, but I do not think your patch is good. If the daemons are running from a previous “cluster”, they will have different X509 certificates loaded and a different state. Instead those daemons should be stopped before proceeding. The changes made in commit 32ba31b should easily apply onto 2.6 as well (you can include a custom patch for the Gentoo ebuild). Since this is not a widespread issue, we will only release the fix in 2.7.

commit 32ba31b
Author: Michael Hanselmann <hansmi@google.com>
Date:   Thu Dec 20 14:22:27 2012 +0100

    Stop all daemons on cluster initialization
    
    Addresses issue 145. When daemons are still running from a previous
    cluster instance, starting the daemons again will fail.
    “ConfigWriter.IsCluster” only looks at the configuration file, which
    could already have been deleted.
    
    Signed-off-by: Michael Hanselmann <hansmi@google.com>
    Reviewed-by: Iustin Pop <iustin@google.com>

```
Originally added on 2012-12-20 13:39:33 +0000 UTC.

Mentioned Issue with GoogleCode ID 145 has been migrated to GitHub with ID #201.
Changed State: Fixed
Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
